### PR TITLE
AP_Scripting: fix PWMSource deletion crash

### PR DESCRIPTION
**Pull request by IamPete1**
This fixes a crash caused by deleting the scripting `PWMSource` object. The object registers a interrupt callback, so when its deleted the callback points to nothing and we crash. 

This moves to handle in a similar way to `i2c` Scripting not the script keeps the object and give the script a pointer to it. This means we can delete it properly if we need to. This does limit the number of `PWMSource` objects to 4, but I think that OK (we can change to more is we need to).

### CODE REVIEW

File: libraries/AP_Scripting/AP_Scripting.cpp [hunk 1]
```diff
@@ -244,6 +244,19 @@ void AP_Scripting::thread(void) {
...
+            if (_pwm_source[i] != nullptr) {
+                // Unregister interrupt and delete
+                // Not that this does not check if the interrupt was successfully detached
+                // That state is not available
+                _pwm_source[i]->set_pin(-1, "");
+                delete _pwm_source[i];
+                _pwm_source[i] = nullptr;
+            }
+        }
+        num_pwm_source = 0;
```
Comment (IamPete1) [inline libraries/AP_Scripting/AP_Scripting.cpp:258]: This is potentially slightly dodgy in that we don't check if the interrupt has been detached. We could just not clear the objects so you would run out faster a few restarts.

Comment (IamPete1) [reply libraries/AP_Scripting/AP_Scripting.cpp:258]: Not deleting would mean that the script tries to register two objects to the same interrupt pin, so presumably it would fail the second time (I have not tested)

File: libraries/AP_Scripting/AP_Scripting.cpp [hunk 2]
```diff
@@ -244,6 +244,19 @@ void AP_Scripting::thread(void) {
         }
         num_i2c_devices = 0;
 
+        // clear allocated PWM sources
+        for (uint8_t i=0; i<SCRIPTING_MAX_NUM_PWM_SOURCE; i++) {
+            if (_pwm_source[i] != nullptr) {
+                // Unregister interrupt and delete
+                // Not that this does not check if the interrupt was successfully detached
+                // That state is not available
+                _pwm_source[i]->set_pin(-1, "");
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Scripting/AP_Scripting.cpp:253]: yes, I think the destructor is a better idea, it can check if interrupt_attached is true and it knows the pin number

File: libraries/AP_Scripting/lua_bindings.cpp [hunk 3]
```diff
@@ -500,6 +500,8 @@ int lua_get_CAN_device2(lua_State *L) {
   directory listing, return table of files in a directory
  */
 int lua_dirlist(lua_State *L) {
+    binding_argcheck(L, 1);
```
Comment (tridge) [inline libraries/AP_Scripting/lua_bindings.cpp:503]: why this change?

File: libraries/AP_Scripting/lua_bindings.cpp [hunk 4]
```diff
@@ -529,8 +531,9 @@ int lua_dirlist(lua_State *L) {
   remove a file
  */
 int lua_removefile(lua_State *L) {
-  const char *filename = luaL_checkstring(L, 1);
-  return luaL_fileresult(L, remove(filename) == 0, filename);
+    binding_argcheck(L, 1);
```
Comment (tridge) [inline libraries/AP_Scripting/lua_bindings.cpp:534]: why this change?

File: libraries/AP_Scripting/lua_bindings.cpp [hunk 5]
```diff
@@ -540,3 +543,24 @@ int SRV_Channels_get_safety_state(lua_State *L) {
     lua_pushboolean(L, data);
     return 1;
 }
+
+int lua_get_PWMSource(lua_State *L) {
+    binding_argcheck(L, 0);
+
+    static_assert(SCRIPTING_MAX_NUM_PWM_SOURCE >= 0, "There cannot be a negative number of PWMSources");
+    if (AP::scripting()->num_pwm_source >= SCRIPTING_MAX_NUM_I2C_DEVICE) {
```
Comment (tridge) [inline libraries/AP_Scripting/lua_bindings.cpp:551]: not I2C (wrong macro)

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/AP_Scripting.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Scripting/AP_Scripting.cpp]: I have added the destructor.

File: libraries/AP_Scripting/lua_bindings.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_Scripting/lua_bindings.cpp:503]: Checks you actually pass 1 argument, so `dirlist('foo', 10)` will now give a error. previously it would have just ignored the extra arguments. Also I'm not exactly sure what `dirlist()` would do now with no arguments. It might try and convert the next thing on the stack to a string. 

The arg check will error out with either "too many arguments" or "too few arguments".

However your right its more of a drive by in this PR. I can split it off into a separate one if you prefer.

**Comment by IamPete1** [comment]
I'm out of the office, so I have not re-tested the latest push on hardware.

File: libraries/AP_Scripting/lua_bindings.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_Scripting/lua_bindings.cpp:503]: `dirlist()` does fail correctly, but you can just pass extra arguments with no error, so this change stops that.

**Comment by IamPete1** [comment]
Retested, still works.

File: libraries/AP_HAL/GPIO.cpp [hunk 6]
```diff
@@ -6,6 +6,14 @@
 
 extern const AP_HAL::HAL& hal;
 
+AP_HAL::PWMSource::~PWMSource()
+{
+    if (interrupt_attached) {
+        // Assume this is always successful
+        hal.gpio->detach_interrupt(_pin);
```
Comment (tridge) [inline libraries/AP_HAL/GPIO.cpp:13]: set interrupt_attached to false, to catch bad re-use

File: libraries/AP_Scripting/lua_bindings.cpp [hunk 7]
```diff
@@ -540,3 +543,24 @@ int SRV_Channels_get_safety_state(lua_State *L) {
     lua_pushboolean(L, data);
     return 1;
 }
+
+int lua_get_PWMSource(lua_State *L) {
+    binding_argcheck(L, 0);
+
```
Comment (tridge) [inline libraries/AP_Scripting/lua_bindings.cpp:549]: use auto *scripting = AP::scripting() to have local var

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Comment by rmackay9** [comment]
We removed this feature from 4.3 so I've removed this from the 4.3 backports project.
